presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcPageSourceFactory.java

```diff
@@ -208,44 +210,71 @@ private static OrcPageSource createOrcPageSource(
-                OrcColumn orcColumn = null;
+                // Get base orc column
+                OrcColumn baseOrcColumn = null;
                 if (useOrcColumnNames) {
-                    orcColumn = fileColumnsByName.get(column.getName().toLowerCase(ENGLISH));
+                    baseOrcColumn = baseFileColumnsByName.get(column.getBaseColumnName().toLowerCase(ENGLISH));
                 }
-                else if (column.getBaseHiveColumnIndex() < fileColumns.size()) {
-                    orcColumn = fileColumns.get(column.getBaseHiveColumnIndex());
+                else if (column.getBaseHiveColumnIndex() < baseFileColumns.size()) {
+                    baseOrcColumn = baseFileColumns.get(column.getBaseHiveColumnIndex());
+                }
+
+                // Get nested orc column and add it to column adaptations
+                OrcColumn nestedOrcColumn = null;
+                if (baseOrcColumn != null) {
+                    nestedOrcColumn = getNestedOrcColumn(baseOrcColumn, column.getHiveColumnProjectionInfo());
                 }
 
                 Type readType = column.getType();
-                if (orcColumn != null) {
+                if (nestedOrcColumn != null) {
                     int sourceIndex = fileReadColumns.size();
                     columnAdaptations.add(ColumnAdaptation.sourceColumn(sourceIndex));
-                    fileReadColumns.add(orcColumn);
+                    fileReadColumns.add(nestedOrcColumn);
                     fileReadTypes.add(readType);
+                }
+                else {
+                    columnAdaptations.add(ColumnAdaptation.nullColumn(readType));
+                }
+            }
+
+            // Build predicates
+            effectivePredicate = effectivePredicate.transform(x -> x.getColumnType() == REGULAR ? x : null);
+            TupleDomainOrcPredicateBuilder predicateBuilder = TupleDomainOrcPredicate.builder()
+                    .setBloomFiltersEnabled(options.isBloomFiltersEnabled());
+            Map<HiveColumnHandle, Domain> effectivePredicateDomains = effectivePredicate.getDomains()
+                    .orElseThrow(() -> new IllegalArgumentException("Effective predicate is none"));
 
+            for (Map.Entry<HiveColumnHandle, Domain> entry : effectivePredicateDomains.entrySet()) {
+                OrcColumn baseOrcColumn = null;
+                HiveColumnHandle column = entry.getKey();
+
+                if (useOrcColumnNames) {
+                    baseOrcColumn = baseFileColumnsByName.get(column.getBaseColumnName().toLowerCase(ENGLISH));
+                }
+                else if (column.getBaseHiveColumnIndex() < baseFileColumns.size()) {
+                    baseOrcColumn = baseFileColumns.get(column.getBaseHiveColumnIndex());
+                }
+
+                if (baseOrcColumn != null) {
+                    OrcColumn orcColumn = getNestedOrcColumn(baseOrcColumn, column.getHiveColumnProjectionInfo());
```

Comment:
@martint I separated the predicate-building from columnAdaptations-building thinking of the possibility that `effectivePredicate` contains a column which is not present in `columns`. 

I don't think this happens right now, because effectivePredicate is still re-tested inside the plan. Can you please confirm that?

The separate computation of predicates would be required if the following state is possible.

```
TableScan
     assignments -->  x := column(a.b)
     effectivePredicate --> a = row(2, 3)        
```

But I don't think this can happen right now, assignments will have column `a` as well, since it's in the filter. Please let me know if we're missing any cases here. Otherwise it should be okay to go with your suggestion.

Reply:
Correct. The predicate has to reference columns listed in the assignments.